### PR TITLE
Feature/rolling in remove dead rooms

### PR DESCRIPTION
**Pull request by lftakakura**
RemoveDeadRooms will not delete all rooms that have not performed ping for a while or have expired occupied timeout. Instead it will segment pods that are in these conditions and do rolling strategy to remove them (creating new ones before removing old).

### CODE REVIEW

File: controller/utils.go [hunk 1]
```diff
@@ -126,6 +127,7 @@ func replacePodsAndWait(
 	wg.Add(len(podsChunk))
 	for _, pod := range podsChunk {
 		go func(pod v1.Pod) {
+			defer GinkgoRecover()
```
Comment (cscatolini) [inline controller/utils.go]: It doesn't seem right to defer test related stuff in here. Why is this necessary?

File: watcher/watcher.go [hunk 2]
```diff
@@ -268,14 +274,17 @@ func (w *Watcher) reportRoomsStatusesRoutine() {
 }
 
 func (w *Watcher) watchRooms() error {
-	w.RemoveDeadRooms()
-	w.AutoScale()
-	w.AddUtilizationMetricsToRedis()
+	l := w.Logger.WithFields(logrus.Fields{
+		"operation": "watcher.watchRooms",
+	})
+	w.WithRedisLock(l, w.AutoScale)
```
Comment (cscatolini) [inline watcher/watcher.go]: Why do we need full redis lock here?

File: watcher/watcher.go [hunk 3]
```diff
@@ -268,14 +274,17 @@ func (w *Watcher) reportRoomsStatusesRoutine() {
...
+	w.WithDownscalingLock(l, w.RemoveDeadRooms)
+	w.WithRedisLock(l, w.AddUtilizationMetricsToRedis)
```
Comment (cscatolini) [inline watcher/watcher.go]: and here?

File: controller/utils.go (hunk 1 above)
Comment (lftakakura) [reply controller/utils.go]: I need this to debug tests. It shouldn't have slipped through. My bad =(

File: watcher/watcher.go (hunk 2 above)
Comment (lftakakura) [reply watcher/watcher.go]: I just kept this and AddUtilizationMetricsToRedis as it was. I'm going to think about using downScaling

File: controller/utils.go (hunk 1 above)
Comment (lftakakura) [reply controller/utils.go]: Removed

**Comment by ghostec** [comment]
Can you add a description of how your changes affect Maestro? :eyes:

File: watcher/watcher.go (hunk 2 above)
Comment (cscatolini) [reply watcher/watcher.go]: ah okay! sorry, with the diff I didn't notice this was already AddUtilizationMetricsToRedis and AutoScale behaviour to lock everything

**Comment by lftakakura** [comment]
> LGTM
> 
> It's up to you if you want to improve AddUtilizationMetricsToRedis and AutoScale locks in this PR or in another one. I'd change it in a separate PR so we can deploy this and validate the behaviour is as expected.

Ok. I'm already working on it. Gonna open a new PR

**Comment by lftakakura** [comment]
> Can you add a description of how your changes affect Maestro? 👀

Sure

File: watcher/watcher.go [hunk 4]
```diff
@@ -472,171 +509,221 @@ func (w *Watcher) WithRedisLock(l *logrus.Entry, f func() error) {
...
+	}
+
+	if len(roomsNoPingSince) > 0 || len(roomsOnOccupiedTimeout) > 0 {
+		podsToReplace, err := w.listPodsToReplace(logger, append(roomsNoPingSince, roomsOnOccupiedTimeout...))
+
+		// load scheduler from database
 		scheduler := models.NewScheduler(w.SchedulerName, "", "")
 		err = scheduler.Load(w.DB)
 		if err != nil {
 			logger.WithError(err).Error("error accessing db while removing dead rooms")
```
Comment (ghostec) [inline watcher/watcher.go]: Right after logging the error, all RemoveDeadRooms does is `logger.Info("finish check of dead rooms")` then it returns nil. Shouldn't you just return this error, remove the else and assume you'll only get there if err == nil?

File: watcher/watcher_test.go [hunk 5]
```diff
@@ -4613,8 +4589,8 @@ var _ = Describe("Watcher", func() {
 				Expect(err).ToNot(HaveOccurred())
 			}
 
-			testing.MockRedisLock(mockRedisClient, downScalingLockKey, config.GetInt("watcher.lockTimeoutMs"), true, nil)
-			testing.MockReturnRedisLock(mockRedisClient, downScalingLockKey, nil)
+			// testing.MockRedisLock(mockRedisClient, downScalingLockKey, config.GetInt("watcher.lockTimeoutMs"), true, nil)
```
Comment (ghostec) [inline watcher/watcher_test.go]: remove?

**Comment by ghostec** [comment]
LGTM
